packages/palette/src/elements/Shelf/Shelf.tsx

```diff
@@ -0,0 +1,289 @@
+
+  const init = () => {
+    if (containerRef.current === null || viewportRef.current === null) return
+
+    // Set page-stops
+    const { current: viewport } = viewportRef
+    const values = cells.map(({ ref }) => ref.current.clientWidth)
+    setPages(paginateCarousel({ viewport: viewport.clientWidth, values }))
+
+    // Set offset to accomodate full-bleed and line up initially with page-margins
+    const { current: container } = containerRef
+    const { x } = container.getBoundingClientRect()
+    setOffset(x)
+    setMounted(true)
+  }
+
+  useEffect(() => {
+    init()
+
+    window.addEventListener("resize", init)
+    return () => {
+      window.removeEventListener("resize", init)
+    }
+  }, [cells])
+
+  const { index: pageIndex, setCursor } = useCursor({
+    max: pages.length,
+  })
+
+  // Keep page index in sync with scroll position
+  useEffect(() => {
+    if (viewportRef.current === null) return
+
+    const { current: viewport } = viewportRef
+
+    const handler = () => {
+      const nearestPage = pages.find((currentPage, i) => {
+        const nextPage = pages[i + 1] ?? Infinity
+        return (
+          viewport.scrollLeft >= currentPage && viewport.scrollLeft < nextPage
+        )
+      })
+
+      setCursor(pages.indexOf(nearestPage))
+
+      const __progress__ = Math.ceil(
+        (viewport.scrollLeft * 100) / (pages[pages.length - 1] - 1)
+      )
+
+      setProgress(__progress__)
+      onProgress && onProgress(progress)
+    }
+
+    viewport.addEventListener("scroll", handler, { passive: true })
+    return () => {
+      viewport.removeEventListener("scroll", handler)
+    }
+  }, [pages])
+
+  // Announce page changes
+  useEffect(() => {
+    onChange && onChange(pageIndex)
+  }, [onChange, pageIndex])
+
+  // Scroll to a specific page-stop
+  const scrollTo = (index: number) => {
+    if (viewportRef.current === null) return
+
+    const { current: viewport } = viewportRef
+    const xPosition = pages[index]
+
+    if (viewport.scrollTo) {
+      viewport.scrollTo({ left: xPosition, behavior: "smooth" })
+      return
+    }
+
+    viewport.scrollLeft = xPosition
+  }
+
+  const handleNext = () => {
+    scrollTo(pageIndex + 1)
+  }
+
+  const handlePrev = () => {
+    scrollTo(pageIndex - 1)
+  }
+
+  return (
+    <Container ref={containerRef as any} {...rest}>
+      <Nav as="nav">
+        <Previous
+          onClick={handlePrev}
+          disabled={pageIndex === 0}
+          aria-label="Previous page"
+        >
+          <ChevronIcon direction="left" width={15} height={15} />
+        </Previous>
+
+        <Next
+          onClick={handleNext}
+          disabled={pageIndex === pages.length - 1}
+          aria-label="Next page"
+        >
+          <ChevronIcon direction="right" width={15} height={15} />
+        </Next>
+      </Nav>
+
+      <FullBleed
+        {...(mounted
+          ? { left: undefined, right: undefined, marginRight: undefined }
```

Comment:
I'd love to figure out a way to do this without JS but I'm not sure it's possible; here we just fake the initial state for SSR — it bleeds off the right edge, but the left is aligned. Once it's mounted we can re-enable the full-bleed then use the offset below as the first element's padding-left and last element's padding-right.

Reply:
This def makes sense! though would suggest moving this block into a variable and adding this as a comment ;)